### PR TITLE
Gradient with Weibull site model

### PR DESCRIPTION
**Pull request by 4ment**
see #230

### CODE REVIEW

File: src/fat_beagle.cpp [hunk 1]
```diff
@@ -79,24 +79,43 @@ double FatBeagle::LogLikelihood(const RootedTree &tree) const {
...
+  EigenMatrixXd dQ = mapQ.replicate(category_count, 1);
+  const EigenVectorXd &category_rates = site_model.GetCategoryRates();
+  for (size_t k = 0; k < category_count; k++) {
+    dQ.row(k) *= category_rates[k];
+  }
+  return dQ;
+}
+
+// Build differential matrix for site model gradient.
+EigenMatrixXd GetSiteModelDifferentialMatrices(SiteModel &site_model,
```
Comment (matsen) [inline src/fat_beagle.cpp:97]: I have a slight preference for `Build` vs `Get` as the latter sounds like we are just pulling out something that's already there.

File: src/fat_beagle.cpp [hunk 2]
```diff
@@ -511,16 +530,41 @@ std::vector<double> RatioGradient(const RootedTree &tree,
...
+      *phylo_model_->GetSiteModel(), *phylo_model_->GetSubstitutionModel());
   auto [log_likelihood, branch_length_gradient] =
-      BranchGradientInternals(tree.Topology(), tree.BranchLengths());
-  // We want the fixed node to have a zero gradient.
-  branch_length_gradient[tree.Topology()->Children()[1]->Id()] = 0.;
+      BranchGradientInternals(tree.Topology(), tree.BranchLengths(), dQ);
 
   std::vector<double> substitution_model_gradient;
   std::vector<double> site_model_gradient;
+  auto site_model = phylo_model_->GetSiteModel();
```
Comment (matsen) [inline src/fat_beagle.cpp:554]: This block and the below can get factored out to reduce code duplication as well, right?

File: src/fat_beagle.hpp [hunk 3]
```diff
@@ -77,7 +77,8 @@ class FatBeagle {
   double LogLikelihoodInternals(const Node::NodePtr topology,
                                 const std::vector<double> &branch_lengths) const;
   std::pair<double, std::vector<double>> BranchGradientInternals(
-      const Node::NodePtr topology, const std::vector<double> &branch_lengths) const;
+      const Node::NodePtr topology, const std::vector<double> &branch_lengths,
```
Comment (matsen) [inline src/fat_beagle.hpp:80]: It'd be nice to have a short comment describing what dQ is.

File: src/rooted_sbn_instance.hpp [hunk 4]
```diff
@@ -213,6 +213,31 @@ TEST_CASE("RootedSBNInstance: clock gradients") {
...
+  for (auto& tree : inst.tree_collection_.trees_) {
+    tree.rates_.assign(tree.rates_.size(), 0.001);
+  }
+  auto param_block_map = inst.GetPhyloModelParamBlockMap();
+  param_block_map.at(WeibullSiteModel::shape_key_).setConstant(0.1);
+
+  auto likelihood = inst.LogLikelihoods();
+  double physher_ll = -4618.2062529058;
+  CHECK_LT(fabs(likelihood[0] - physher_ll), 0.0001);
+  std::cout << likelihood[0] << std::endl;
```
Comment (matsen) [inline src/rooted_sbn_instance.hpp:232]: Trim this output?

File: src/site_model.cpp [hunk 5]
```diff
@@ -34,16 +34,29 @@ void WeibullSiteModel::SetParameters(const EigenVectorXdRef param_vector) {
...
+  double mean_rate = 0;
+  double mean_rate_derivative = 0;
+  std::vector<double> deriv_unscaled_rates(category_count_);
   for (size_t i = 0; i < category_count_; i++) {
     double quantile = (2.0 * i + 1.0) / (2.0 * category_count_);
     // Set rate to inverse CDF at quantile.
     category_rates_[i] = pow(-std::log(1.0 - quantile), 1.0 / shape_);
-    mean += category_rates_[i];
+    mean_rate += category_rates_[i];
+    // Derivative of unormalized rate i wrt shape
```
Comment (matsen) [inline src/site_model.cpp:45]: This is awesome! Do you have any helpful links for understanding the derivation?

Comment (4ment) [reply src/site_model.cpp:45]: It is coming from physher and there is even less comments in the code

**Pull request merged**